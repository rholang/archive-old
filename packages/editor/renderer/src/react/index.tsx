import * as React from 'react';
// @ts-ignore: unused variable
// prettier-ignore
import { ComponentType, Consumer, Provider } from 'react';
import { Fragment, Mark, MarkType, Node, Schema } from 'prosemirror-model';
import { Serializer } from '../';
import { RendererAppearance } from '../ui/Renderer/types';
import { AnalyticsEventPayload } from '../analytics/events';

import {
  Doc,
  mergeTextNodes,
  isTextWrapper,
  TextWrapper,
  toReact,
} from './nodes';

import { toReact as markToReact } from './marks';
import {
  ProviderFactory,
  getMarksByOrder,
  isSameMark,
  EventHandlers,
  ExtensionHandlers,
  calcTableColumnWidths,
} from '@atlaskit/editor-common';
import { getText } from '../utils';

export interface RendererContext {
  objectAri?: string;
  containerAri?: string;
  adDoc?: any;
  schema?: Schema;
}

export interface ConstructorParams {
  providers?: ProviderFactory;
  eventHandlers?: EventHandlers;
  extensionHandlers?: ExtensionHandlers;
  portal?: HTMLElement;
  objectContext?: RendererContext;
  appearance?: RendererAppearance;
  disableHeadingIDs?: boolean;
  allowDynamicTextSizing?: boolean;
  allowHeadingAnchorLinks?: boolean;
  allowColumnSorting?: boolean;
  fireAnalyticsEvent?: (event: AnalyticsEventPayload) => void;
  shouldOpenMediaViewer?: boolean;
}

type MarkWithContent = Partial<Mark<any>> & {
  content: Array<MarkWithContent | Node<any>>;
};

function mergeMarks(marksAndNodes: Array<MarkWithContent | Node>) {
  return marksAndNodes.reduce(
    (acc, markOrNode) => {
      const prev = (acc.length && acc[acc.length - 1]) || null;

      if (
        markOrNode.type instanceof MarkType &&
        prev &&
        prev.type instanceof MarkType &&
        Array.isArray(prev.content) &&
        isSameMark(prev as Mark, markOrNode as Mark)
      ) {
        prev.content = mergeMarks(
          prev.content.concat((markOrNode as MarkWithContent).content),
        );
      } else {
        acc.push(markOrNode);
      }

      return acc;
    },
    [] as Array<MarkWithContent | Node>,
  );
}

export default class ReactSerializer implements Serializer<JSX.Element> {
  private providers?: ProviderFactory;
  private eventHandlers?: EventHandlers;
  private extensionHandlers?: ExtensionHandlers;
  private portal?: HTMLElement;
  private rendererContext?: RendererContext;
  private appearance?: RendererAppearance;
  private disableHeadingIDs?: boolean;
  private headingIds: string[] = [];
  private allowDynamicTextSizing?: boolean;
  private allowHeadingAnchorLinks?: boolean;
  private allowColumnSorting?: boolean;
  private fireAnalyticsEvent?: (event: AnalyticsEventPayload) => void;
  private shouldOpenMediaViewer?: boolean;

  constructor({
    providers,
    eventHandlers,
    extensionHandlers,
    portal,
    objectContext,
    appearance,
    disableHeadingIDs,
    allowDynamicTextSizing,
    allowHeadingAnchorLinks,
    allowColumnSorting,
    fireAnalyticsEvent,
    shouldOpenMediaViewer,
  }: ConstructorParams) {
    this.providers = providers;
    this.eventHandlers = eventHandlers;
    this.extensionHandlers = extensionHandlers;
    this.portal = portal;
    this.rendererContext = objectContext;
    this.appearance = appearance;
    this.disableHeadingIDs = disableHeadingIDs;
    this.allowDynamicTextSizing = allowDynamicTextSizing;
    this.allowHeadingAnchorLinks = allowHeadingAnchorLinks;
    this.allowColumnSorting = allowColumnSorting;
    this.fireAnalyticsEvent = fireAnalyticsEvent;
    this.shouldOpenMediaViewer = shouldOpenMediaViewer;
  }

  private resetState() {
    this.headingIds = [];
  }

  serializeFragment(
    fragment: Fragment,
    props: any = {},
    target: any = Doc,
    key: string = 'root-0',
    parentInfo?: { parentIsIncompleteTask: boolean; path: Array<Node> },
  ): JSX.Element | null {
    // This makes sure that we reset internal state on re-render.
    if (key === 'root-0') {
      this.resetState();
    }

    const content = ReactSerializer.getChildNodes(fragment).map(
      (node, index) => {
        if (isTextWrapper(node)) {
          return this.serializeTextWrapper(node.content);
        }

        let props;

        if (node.type.name === 'table') {
          props = this.getTableProps(node);
        } else if (node.type.name === 'date') {
          props = this.getDateProps(node, parentInfo);
        } else if (node.type.name === 'heading') {
          props = this.getHeadingProps(node, parentInfo && parentInfo.path);
        } else if (['tableHeader', 'tableRow'].indexOf(node.type.name) > -1) {
          props = this.getTableChildrenProps(node);
        } else if (node.type.name === 'media') {
          props = this.getMediaProps(node);
        } else {
          props = this.getProps(node);
        }

        let currentPath = (parentInfo && parentInfo.path) || [];

        const parentIsIncompleteTask =
          node.type.name === 'taskItem' && node.attrs.state !== 'DONE';

        let pInfo = {
          parentIsIncompleteTask,
          path: [...currentPath, node],
        };

        const serializedContent = this.serializeFragment(
          node.content,
          props,
          toReact(node),
          `${node.type.name}-${index}`,
          pInfo,
        );

        if (node.marks && node.marks.length) {
          return ([] as Array<Mark>)
            .concat(node.marks)
            .reverse()
            .reduce((acc, mark) => {
              return this.renderMark(
                markToReact(mark),
                this.getMarkProps(mark),
                `${mark.type.name}-${index}`,
                acc,
              );
            }, serializedContent);
        }

        return serializedContent;
      },
    );

    return this.renderNode(target, props, key, content);
  }

  private serializeTextWrapper(content: Node[]) {
    return ReactSerializer.buildMarkStructure(content).map((mark, index) =>
      this.serializeMark(mark, index),
    );
  }

  private serializeMark(mark: Mark, index: number = 0) {
    if (mark.type.name === 'text') {
      return (mark as any).text;
    }

    const content = ((mark as any).content || []).map(
      (child: Mark, index: number) => this.serializeMark(child, index),
    );
    return this.renderMark(
      markToReact(mark),
      this.getMarkProps(mark),
      `${mark.type.name}-${index}`,
      content,
    );
  }

  private renderNode(
    NodeComponent: ComponentType<any>,
    props: any,
    key: string,
    content: string | JSX.Element | any[] | null | undefined,
  ): JSX.Element {
    return (
      <NodeComponent key={key} {...props}>
        {content}
      </NodeComponent>
    );
  }

  private renderMark(
    MarkComponent: ComponentType<any>,
    props: any,
    key: string,
    content: any,
  ) {
    return (
      <MarkComponent key={key} {...props}>
        {content}
      </MarkComponent>
    );
  }

  private getTableChildrenProps(node: Node) {
    return {
      ...this.getProps(node),
      allowColumnSorting: this.allowColumnSorting,
    };
  }

  private getTableProps(node: Node) {
    return {
      ...this.getProps(node),
      allowColumnSorting: this.allowColumnSorting,
      columnWidths: calcTableColumnWidths(node),
      tableNode: node,
    };
  }

  private getDateProps(
    node: Node,
    parentInfo: { parentIsIncompleteTask: boolean } | undefined,
  ) {
    return {
      timestamp: node.attrs && node.attrs.timestamp,
      parentIsIncompleteTask: parentInfo && parentInfo.parentIsIncompleteTask,
    };
  }
  private getMediaProps(node: Node) {
    return {
      ...this.getProps(node),
      shouldOpenMediaViewer: this.shouldOpenMediaViewer,
    };
  }

  private getProps(node: Node) {
    return {
      text: node.text,
      providers: this.providers,
      eventHandlers: this.eventHandlers,
      extensionHandlers: this.extensionHandlers,
      portal: this.portal,
      rendererContext: this.rendererContext,
      serializer: this,
      content: node.content ? node.content.toJSON() : undefined,
      allowDynamicTextSizing: this.allowDynamicTextSizing,
      allowHeadingAnchorLinks: this.allowHeadingAnchorLinks,
      rendererAppearance: this.appearance,
      fireAnalyticsEvent: this.fireAnalyticsEvent,
      ...node.attrs,
    };
  }

  private headingAnchorSupported(path: Array<Node> = []): boolean {
    return (
      path.length === 0 || path[path.length - 1].type.name === 'layoutColumn'
    );
  }

  private getHeadingProps(node: Node, path: Array<Node> = []) {
    return {
      ...node.attrs,
      content: node.content ? node.content.toJSON() : undefined,
      headingId: this.getHeadingId(node),
      showAnchorLink:
        this.allowHeadingAnchorLinks &&
        !this.disableHeadingIDs &&
        this.headingAnchorSupported(path),
    };
  }

  // The return value of this function is NOT url encoded,
  // In HTML5 standard, id can contain any characters, encoding is no necessary.
  // Plus we trying to avoid double encoding, therefore we leave the value as is.
  // Remember to use encodeURIComponent when generating url from the id value.
  private getHeadingId(node: Node) {
    if (this.disableHeadingIDs || !node.content.size) {
      return;
    }

    // We are not use node.textContent here, because we would like to handle cases where
    // headings only contain inline blocks like emoji, status and date.
    const nodeContent = (node as any).content
      .toJSON()
      .reduce((acc: string, node: any) => acc.concat(getText(node) || ''), '')
      .trim()
      .replace(/\s/g, '-');

    if (!nodeContent) {
      return;
    }

    return this.getUniqueHeadingId(nodeContent);
  }

  private getUniqueHeadingId(baseId: string, counter = 0): string {
    if (counter === 0 && this.headingIds.indexOf(baseId) === -1) {
      this.headingIds.push(baseId);
      return baseId;
    } else if (counter !== 0) {
      const headingId = `${baseId}.${counter}`;
      if (this.headingIds.indexOf(headingId) === -1) {
        this.headingIds.push(headingId);
        return headingId;
      }
    }

    return this.getUniqueHeadingId(baseId, ++counter);
  }

  private getMarkProps(mark: Mark): any {
    const { key, ...otherAttrs } = mark.attrs;
    return {
      eventHandlers: this.eventHandlers,
      markKey: key,
      ...otherAttrs,
    };
  }

  static getChildNodes(fragment: Fragment): (Node | TextWrapper)[] {
    const children: Node[] = [];
    fragment.forEach(node => {
      children.push(node);
    });
    return mergeTextNodes(children) as Node[];
  }

  static getMarks(node: Node): Mark[] {
    if (!node.marks || node.marks.length === 0) {
      return [];
    }

    return getMarksByOrder(node.marks);
  }

  static buildMarkStructure(content: Node[]) {
    return mergeMarks(
      content.map(node => {
        const nodeMarks = this.getMarks(node);
        if (nodeMarks.length === 0) {
          return node;
        }

        return nodeMarks.reverse().reduce(
          (acc, mark) => {
            const { eq } = mark;

            return {
              ...mark,
              eq,
              content: [acc],
            };
          },
          node as any,
        );
      }),
    ) as Mark[];
  }

  static fromSchema(
    _schema: Schema,
    {
      providers,
      eventHandlers,
      extensionHandlers,
      appearance,
      disableHeadingIDs,
      allowDynamicTextSizing,
      allowHeadingAnchorLinks,
      allowColumnSorting,
    }: ConstructorParams,
  ): ReactSerializer {
    // TODO: Do we actually need the schema here?
    return new ReactSerializer({
      providers,
      eventHandlers,
      extensionHandlers,
      appearance,
      disableHeadingIDs,
      allowDynamicTextSizing,
      allowHeadingAnchorLinks,
      allowColumnSorting,
    });
  }
}
