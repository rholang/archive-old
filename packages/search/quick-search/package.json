{
  "name": "@atlaskit/quick-search",
  "version": "7.8.1",
  "description": "A quick-search component",
  "license": "Apache-2.0",
  "module": "dist/esm/index.js",
  "main": "dist/cjs/index.js",
  "types": "dist/cjs/index.d.ts",
  "atlaskit:src": "src/index.ts",
  "author": "Atlassian Pty Ltd",
  "maintainers": [
    "Ahmed Fouad",
    "Patrick Teen"
  ],
  "atlaskit": {
    "name": "Quick Search",
    "team": "search"
  },
  "repository": "https://bitbucket.org/atlassian/atlaskit-mk-2",
  "dependencies": {
    "@atlaskit/analytics": "^6.1.1",
    "@atlaskit/avatar": "^17.1.3",
    "@atlaskit/icon": "^19.0.6",
    "@atlaskit/item": "^10.1.6",
    "@atlaskit/spinner": "^12.1.1",
    "@atlaskit/theme": "^9.2.4",
    "@babel/runtime": "^7.0.0"
  },
  "peerDependencies": {
    "react": "^16.8.0",
    "styled-components": "^3.2.6"
  },
  "devDependencies": {
    "@atlaskit/avatar": "^17.1.3",
    "@atlaskit/docs": "^8.1.8",
    "@atlaskit/drawer": "^5.0.13",
    "@atlaskit/navigation": "^35.2.5",
    "enzyme": "^3.7.0",
    "prop-types": "^15.5.10",
    "react": "^16.8.0",
    "sinon": "^2.2.0",
    "uuid": "^3.1.0"
  }
}