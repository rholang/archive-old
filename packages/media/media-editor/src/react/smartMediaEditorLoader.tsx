import * as React from 'react';
import { ModalSpinner } from '@atlaskit/media-ui';
import { WithMediaClientConfigProps } from '@atlaskit/media-client';
import { N700A } from '@atlaskit/theme/colors';
import { SmartMediaEditorProps } from './smartMediaEditor';

type SmartEditorWithMediaClientConfigProps = WithMediaClientConfigProps<
  SmartMediaEditorProps
>;
type SmartEditorWithMediaClientConfigComponent = React.ComponentType<
  SmartEditorWithMediaClientConfigProps
>;

interface AsyncSmartMediaEditorState {
  SmartMediaEditor?: SmartEditorWithMediaClientConfigComponent;
}

export default class AsyncSmartMediaEditor extends React.PureComponent<
  SmartEditorWithMediaClientConfigProps & AsyncSmartMediaEditorState,
  AsyncSmartMediaEditorState & { isErrored: boolean }
> {
  static displayName = 'AsyncSmartMediaEditor';
  static SmartMediaEditor?: SmartEditorWithMediaClientConfigComponent;

  state = {
    // Set state value to equal to current static value of this class.
    SmartMediaEditor: AsyncSmartMediaEditor.SmartMediaEditor,
    isErrored: false,
  };

  async UNSAFE_componentWillMount() {
    if (!this.state.SmartMediaEditor) {
      try {
        const [mediaClient, smartEditorModule] = await Promise.all([
          import(/* webpackChunkName:"@atlaskit-media-client" */ '@atlaskit/media-client'),
          import(/* webpackChunkName:"@atlaskit-internal_smart-media-editor" */ './smartMediaEditor'),
        ]);
        AsyncSmartMediaEditor.SmartMediaEditor = mediaClient.withMediaClient(
          smartEditorModule.default,
        );
        this.setState({
          SmartMediaEditor: AsyncSmartMediaEditor.SmartMediaEditor,
        });
      } catch (e) {
        // eslint-disable-next-line no-console
        console.error(e);
        this.setState({ isErrored: true });
      }
    }
  }

  render() {
    const { isErrored } = this.state;

    if (isErrored) {
      return null;
    }
    if (!this.state.SmartMediaEditor) {
      return <ModalSpinner blankedColor={N700A} invertSpinnerColor={true} />;
    }

    return <this.state.SmartMediaEditor {...this.props} />;
  }
}
