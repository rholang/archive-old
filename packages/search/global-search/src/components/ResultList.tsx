import * as React from 'react';
import styled from 'styled-components';
import { colors } from '@atlaskit/theme';
import SearchIcon from '@atlaskit/icon/glyph/search';
import {
  ObjectResult as ObjectResultComponent,
  PersonResult as PersonResultComponent,
  ContainerResult as ContainerResultComponent,
  ResultBase,
} from '@atlaskit/quick-search';
import { FormattedMessage } from 'react-intl';
import { messages } from '../messages';
import {
  Result,
  ContainerResult,
  JiraResult,
  PersonResult,
  ResultType,
  ConfluenceObjectResult,
  JiraProjectType,
  ContentType,
  AnalyticsType,
} from '../model/Result';
import { SelectedIcon } from './styled';
import { getAvatarForConfluenceObjectResult } from '../util/confluence-avatar-util';
import { getDefaultAvatar } from '../util/jira-avatar-util';
import DarkReturn from '../assets/DarkReturn';
import Return from '../assets/Return';
import { injectFeatures } from './FeaturesProvider';
import { ConfluenceFeatures, JiraFeatures } from '../util/features';

export interface Props {
  results: Result[];
  sectionIndex: number;
  analyticsData?: {};
  features: ConfluenceFeatures | JiraFeatures; // this component is shared
}

const extractAvatarData = (jiraResult: JiraResult) =>
  jiraResult.avatarUrl
    ? { avatarUrl: jiraResult.avatarUrl }
    : {
        avatar: getDefaultAvatar(jiraResult.contentType),
      };

const selectedIcon = (
  <SelectedIcon>
    <DarkReturn />
  </SelectedIcon>
);

const getI18nJiraContainerName = (
  projectType: JiraProjectType,
): JSX.Element | undefined => {
  switch (projectType) {
    case JiraProjectType.Business: {
      return (
        <FormattedMessage {...messages.jira_project_type_business_project} />
      );
    }
    case JiraProjectType.Software: {
      return (
        <FormattedMessage {...messages.jira_project_type_software_project} />
      );
    }
    case JiraProjectType.ServiceDesk: {
      return (
        <FormattedMessage
          {...messages.jira_project_type_service_desk_project}
        />
      );
    }
    case JiraProjectType.Ops: {
      return <FormattedMessage {...messages.jira_project_type_ops_project} />;
    }
  }
};

const getI18nJiraContentType = (
  contentType: ContentType,
): JSX.Element | undefined => {
  switch (contentType) {
    case ContentType.JiraBoard: {
      return <FormattedMessage {...messages.jira_result_type_board} />;
    }
    case ContentType.JiraFilter: {
      return <FormattedMessage {...messages.jira_result_type_filter} />;
    }
  }
  return undefined;
};

// Being tested as part of the 'complex' experiment, to improve scannability when there
// is lots of text.
const LightSubtextWrapper = styled.span`
  color: ${colors.N90};
`;

const getI18nConfluenceContainerSubtext = (
  containerName: string,
  friendlyLastModified: string | undefined,
) => {
  const containerText = friendlyLastModified ? (
    <FormattedMessage
      {...messages.confluence_container_subtext_with_modified_date}
      values={{
        containerName,
        friendlyLastModified,
      }}
    />
  ) : (
    containerName
  );

  return <LightSubtextWrapper>{containerText}</LightSubtextWrapper>;
};

export const getUniqueResultId = (result: Result): string =>
  result.key ? result.key : `${result.contentType}-${result.resultId}`;

class ResultList extends React.Component<Props> {
  render() {
    const { results, sectionIndex, features } = this.props;

    return results.map((result, index) => {
      const resultType: ResultType = result.resultType;
      const analyticsData = {
        sectionIndex,
        indexWithinSection: index,
        containerId: result.containerId,
        experimentId: result.experimentId,
        ...this.props.analyticsData,
        contentType: result.contentType,
        resultId: result.resultId,
        isRecentResult: !!result.isRecentResult,
      };

      // Make sure that key and resultId are unique across all search results
      const uniqueResultId = getUniqueResultId(result);
      const uniqueKey = `${uniqueResultId}_${Date.now()}`; // same result might appear on two successive search

      switch (resultType) {
        case ResultType.ConfluenceObjectResult: {
          const confluenceResult = result as ConfluenceObjectResult;

          const subText = getI18nConfluenceContainerSubtext(
            confluenceResult.containerName,
            confluenceResult.friendlyLastModified,
          );

          return (
            <ObjectResultComponent
              key={uniqueKey}
              resultId={uniqueResultId}
              name={confluenceResult.name}
              href={confluenceResult.href}
              type={confluenceResult.analyticsType}
              containerName={subText}
              avatar={getAvatarForConfluenceObjectResult(confluenceResult)}
              analyticsData={analyticsData}
              selectedIcon={selectedIcon}
            />
          );
        }
        case ResultType.JiraProjectResult: {
          const jiraResult = result as JiraResult;
          const avatarData = extractAvatarData(jiraResult);

          const containerNameElement = jiraResult.projectType
            ? getI18nJiraContainerName(jiraResult.projectType)
            : null;

          return (
            <ContainerResultComponent
              key={uniqueKey}
              resultId={uniqueResultId}
              name={jiraResult.name}
              href={jiraResult.href}
              type={jiraResult.analyticsType}
              subText={containerNameElement}
              {...avatarData}
              analyticsData={analyticsData}
              selectedIcon={selectedIcon}
            />
          );
        }
        case ResultType.JiraObjectResult: {
          const jiraResult = result as JiraResult;
          const avatarData = extractAvatarData(jiraResult);

          const objectKey =
            jiraResult.contentType === 'jira-board' ||
            jiraResult.contentType === 'jira-filter'
              ? getI18nJiraContentType(jiraResult.contentType)
              : jiraResult.objectKey;

          return (
            <ObjectResultComponent
              key={uniqueKey}
              resultId={uniqueResultId}
              name={jiraResult.name}
              href={jiraResult.href}
              type={jiraResult.analyticsType}
              objectKey={objectKey}
              containerName={jiraResult.containerName}
              {...avatarData}
              analyticsData={analyticsData}
              selectedIcon={selectedIcon}
            />
          );
        }
        case ResultType.GenericContainerResult: {
          const containerResult = result as ContainerResult;
          return (
            <ContainerResultComponent
              key={uniqueKey}
              resultId={uniqueResultId}
              name={containerResult.name}
              href={containerResult.href}
              type={containerResult.analyticsType}
              avatarUrl={containerResult.avatarUrl}
              analyticsData={analyticsData}
              selectedIcon={selectedIcon}
            />
          );
        }
        case ResultType.PersonResult: {
          const personResult = result as PersonResult;

          const presenceMessage = features.complexSearchExtensionsEnabled ? (
            <LightSubtextWrapper>
              {personResult.presenceMessage}
            </LightSubtextWrapper>
          ) : (
            personResult.presenceMessage
          );

          return (
            <PersonResultComponent
              key={uniqueKey}
              resultId={uniqueResultId}
              name={personResult.name}
              href={personResult.href}
              type={personResult.analyticsType}
              avatarUrl={personResult.avatarUrl}
              mentionName={personResult.mentionName}
              presenceMessage={presenceMessage}
              analyticsData={analyticsData}
              selectedIcon={selectedIcon}
              target="_blank"
            />
          );
        }

        case ResultType.JiraIssueAdvancedSearch: {
          return (
            <ResultBase
              href={result.href}
              resultId="jira-advanced-issue-search"
              text={<FormattedMessage {...messages.jira_view_all_issues} />}
              icon={<SearchIcon size="medium" label="View all issues" />}
              type={AnalyticsType.LinkPostQueryAdvancedSearchJira}
              key={uniqueKey}
              elemAfter={<Return />}
            />
          );
        }
        default: {
          // Make the TS compiler verify that all enums have been matched
          const _nonExhaustiveMatch: never = resultType;
          throw new Error(
            `Non-exhaustive match for result type: ${_nonExhaustiveMatch}`,
          );
        }
      }
    });
  }
}

export const UnwrappedResultList = ResultList;

export default injectFeatures(ResultList);
