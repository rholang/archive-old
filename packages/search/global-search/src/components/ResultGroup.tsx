import * as React from 'react';
import styled from 'styled-components';
import { gridSize } from '@atlaskit/theme';
import { ResultItemGroup, CancelableEvent } from '@atlaskit/quick-search';
import Badge from '@atlaskit/badge';

import { Result } from '../model/Result';
import ResultList from './ResultList';
import { injectIntl, InjectedIntlProps } from 'react-intl';
import ShowMoreButton from './ShowMoreButton';

export interface Props {
  title?: JSX.Element | string;
  results: Result[];
  sectionIndex: number;
  analyticsData?: {};
  showTotalSize: boolean;
  totalSize: number;
  showMoreButton: boolean;
  onShowMoreClicked: () => void;
  onSearchMoreAdvancedSearch: undefined | ((e: CancelableEvent) => void);
  query: string;
}

const TitlelessGroupWrapper = styled.div`
  margin-top: ${gridSize() * 1.5}px;
`;

const BadgeContainer = styled.span`
  margin-left: ${gridSize()}px;
`;

export class ResultGroup extends React.Component<Props & InjectedIntlProps> {
  render() {
    const {
      title,
      results,
      sectionIndex,
      showTotalSize,
      totalSize,
      showMoreButton,
      onShowMoreClicked,
      onSearchMoreAdvancedSearch,
      query,
    } = this.props;

    if (results.length === 0) {
      return null;
    }

    const moreButton = showMoreButton ? (
      <ShowMoreButton
        resultLength={results.length}
        onShowMoreClicked={onShowMoreClicked}
        onSearchMoreAdvancedSearch={onSearchMoreAdvancedSearch}
        totalSize={totalSize}
        query={query}
        // this will force new show more button every click show more to fix scrolling
        key={`show_more_${results.length}`}
      />
    ) : null;

    if (!title) {
      return (
        <TitlelessGroupWrapper>
          <ResultList
            analyticsData={this.props.analyticsData}
            results={results}
            sectionIndex={sectionIndex}
          />
          {moreButton}
        </TitlelessGroupWrapper>
      );
    }

    const titleView = showTotalSize ? (
      <>
        <span>{title}</span>
        <BadgeContainer>
          <Badge max={99}>{totalSize}</Badge>
        </BadgeContainer>
      </>
    ) : (
      <span>{title}</span>
    );

    return (
      <ResultItemGroup title={titleView}>
        <ResultList
          analyticsData={this.props.analyticsData}
          results={results}
          sectionIndex={sectionIndex}
        />
        {moreButton}
      </ResultItemGroup>
    );
  }
}

export default injectIntl(ResultGroup);
