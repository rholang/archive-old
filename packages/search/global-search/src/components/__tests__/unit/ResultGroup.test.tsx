import * as React from 'react';

import { shallow } from 'enzyme';
import { ResultGroup, Props } from '../../ResultGroup';
import { makeConfluenceObjectResult } from '../../../__tests__/unit/_test-util';
import ResultList from '../../ResultList';
import { ResultItemGroup } from '@atlaskit/quick-search';

const mockedIntl: ReactIntl.InjectedIntl = {} as ReactIntl.InjectedIntl;

function render(partialProps: Partial<Props>) {
  const props: Props = {
    title: 'some title',
    results: [],
    sectionIndex: 0,
    totalSize: 0,
    showTotalSize: false,
    showMoreButton: false,
    onSearchMoreAdvancedSearch: undefined,
    onShowMoreClicked: () => {},
    query: '',
    ...partialProps,
  };

  return shallow(<ResultGroup {...props} intl={mockedIntl} />);
}

describe('<ResultGroup />', () => {
  it('should render nothing when there are no results', () => {
    const wrapper = render({ results: [] });

    expect(wrapper.isEmptyRender()).toBe(true);
  });

  it('should render a result list when there are results', () => {
    const wrapper = render({ results: [makeConfluenceObjectResult()] });

    expect(wrapper.find(ResultList).exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a result list even with no titles', () => {
    const wrapper = render({
      results: [makeConfluenceObjectResult()],
      title: '',
    });

    expect(wrapper.find(ResultItemGroup)).toHaveLength(0);
    expect(wrapper.find(ResultList).exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render total if showTotalSize is true but there is no title', () => {
    const wrapper = render({
      results: [makeConfluenceObjectResult()],
      showTotalSize: true,
      totalSize: 9999,
      title: '',
    });

    expect(wrapper.find(ResultItemGroup)).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render total if showTotalSize is false and there is a title', () => {
    const wrapper = render({
      results: [makeConfluenceObjectResult()],
      showTotalSize: false,
      totalSize: 9999,
      title: 'Some title',
    });

    const title = wrapper.find(ResultItemGroup).prop('title');
    expect(title).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render total if showTotalSize is true and there is a title', () => {
    const wrapper = render({
      results: [makeConfluenceObjectResult()],
      showTotalSize: true,
      totalSize: 9999,
      title: 'Some title',
    });

    const title = wrapper.find(ResultItemGroup).prop('title');
    expect(title).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});
