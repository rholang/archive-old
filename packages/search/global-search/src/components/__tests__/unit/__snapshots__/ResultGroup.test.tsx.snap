// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultGroup /> should not render total if showTotalSize is false and there is a title 1`] = `
<ResultItemGroup
  title={
    <span>
      Some title
    </span>
  }
>
  <Component
    results={
      Array [
        Object {
          "analyticsType": "result-confluence",
          "avatarUrl": "avatarUrl",
          "containerId": "containerId",
          "containerName": "containerName",
          "contentType": "confluence-page",
          "friendlyLastModified": "friendly-last-modified",
          "href": "href",
          "name": "name",
          "resultId": "resultId",
          "resultType": "confluence-object-result",
        },
      ]
    }
    sectionIndex={0}
  />
</ResultItemGroup>
`;

exports[`<ResultGroup /> should not render total if showTotalSize is true but there is no title 1`] = `
<styled.div>
  <Component
    results={
      Array [
        Object {
          "analyticsType": "result-confluence",
          "avatarUrl": "avatarUrl",
          "containerId": "containerId",
          "containerName": "containerName",
          "contentType": "confluence-page",
          "friendlyLastModified": "friendly-last-modified",
          "href": "href",
          "name": "name",
          "resultId": "resultId",
          "resultType": "confluence-object-result",
        },
      ]
    }
    sectionIndex={0}
  />
</styled.div>
`;

exports[`<ResultGroup /> should render a result list even with no titles 1`] = `
<styled.div>
  <Component
    results={
      Array [
        Object {
          "analyticsType": "result-confluence",
          "avatarUrl": "avatarUrl",
          "containerId": "containerId",
          "containerName": "containerName",
          "contentType": "confluence-page",
          "friendlyLastModified": "friendly-last-modified",
          "href": "href",
          "name": "name",
          "resultId": "resultId",
          "resultType": "confluence-object-result",
        },
      ]
    }
    sectionIndex={0}
  />
</styled.div>
`;

exports[`<ResultGroup /> should render a result list when there are results 1`] = `
<ResultItemGroup
  title={
    <span>
      some title
    </span>
  }
>
  <Component
    results={
      Array [
        Object {
          "analyticsType": "result-confluence",
          "avatarUrl": "avatarUrl",
          "containerId": "containerId",
          "containerName": "containerName",
          "contentType": "confluence-page",
          "friendlyLastModified": "friendly-last-modified",
          "href": "href",
          "name": "name",
          "resultId": "resultId",
          "resultType": "confluence-object-result",
        },
      ]
    }
    sectionIndex={0}
  />
</ResultItemGroup>
`;

exports[`<ResultGroup /> should render total if showTotalSize is true and there is a title 1`] = `
<ResultItemGroup
  title={
    <React.Fragment>
      <span>
        Some title
      </span>
      <styled.span>
        <Badge
          max={99}
        >
          9999
        </Badge>
      </styled.span>
    </React.Fragment>
  }
>
  <Component
    results={
      Array [
        Object {
          "analyticsType": "result-confluence",
          "avatarUrl": "avatarUrl",
          "containerId": "containerId",
          "containerName": "containerName",
          "contentType": "confluence-page",
          "friendlyLastModified": "friendly-last-modified",
          "href": "href",
          "name": "name",
          "resultId": "resultId",
          "resultType": "confluence-object-result",
        },
      ]
    }
    sectionIndex={0}
  />
</ResultItemGroup>
`;
